src/compile.rs

```diff
@@ -459,6 +459,197 @@ pub fn validate_rules(statements: &BTreeSet<ValidatedStatement>) -> Result<(), C
+
+                // If we have found one associated resource
+                // we can continue
+                continue;
+            }
+        }
+    }
+}
+
+// Go through all of the functions and check if they are castable
+// base on functions they call.
+pub fn determine_castable(functions: &mut FunctionMap, types: &TypeMap) -> u64 {
+    let mut num_changed: u64 = 0;
+    // We need tmp_functions to avoid a immutable borrow after a mutable one.
+    let tmp_functions = functions.clone();
+    'outer: for func in functions.values_mut() {
+        // If we are already false there is no reason to check our called functions.
+        if !func.is_castable {
+            continue;
+        }
+        for call in func.original_body {
+            if let Statement::Call(call) = call {
+                if let Some(inner_func) = tmp_functions.get(&call.get_cil_name()) {
+                    if !inner_func.is_castable {
+                        num_changed += 1;
+                        func.is_castable = false;
+                        continue 'outer;
+                    }
+                // The call may be a built-in so check the args
+                } else {
+                    for arg in &call.args {
+                        if let Argument::Var(arg) = &arg.0 {
+                            // Need to special case this.*
+                            if arg.to_string().contains("this.") {
+                                num_changed += 1;
+                                func.is_castable = false;
+                                continue 'outer;
+                            }
+                            if let Some(ti) = types.get(arg.as_ref()) {
+                                if ti.is_associated_resource(types) {
+                                    num_changed += 1;
+                                    func.is_castable = false;
+                                    continue 'outer;
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+    num_changed
+}
+
+pub fn initialize_terminated<'a>(
+    functions: &'a FunctionMap<'a>,
+) -> (Vec<String>, Vec<FunctionInfo<'a>>) {
+    let mut term_ret_vec: Vec<String> = Vec::new();
+    let mut nonterm_ret_vec: Vec<FunctionInfo> = Vec::new();
+
+    for func in functions.values() {
+        let mut is_term = true;
+
+        for call in func.original_body {
+            if let Statement::Call(call) = call {
+                match call.check_builtin() {
+                    Some(_) => {
+                        continue;
+                    }
+                    None => {
+                        is_term = false;
+                        break;
+                    }
+                }
+            }
+        }
+        if is_term {
+            term_ret_vec.push(func.get_cil_name().clone());
+        } else {
+            nonterm_ret_vec.push(func.clone());
```

Comment:
Why are we saving the full FunctionInfo on this one instead of just the name?

Reply:
We need the original_body to look through the statements and this saves us from having to go back to the function map.

Reply:
Well, this is another occasion where performance testing would help, but I'd guess that the map lookups are more performant than cloning all the FIs.  If that's the only tradeoff, I'd prefer to save the names and look them up in the map again.